src/extension/infer.rs

```diff
@@ -663,31 +676,98 @@ impl UnificationContext {
         self.results()
     }
 
-    /// Instantiate all variables in the graph with the empty extension set.
+    /// Gather all the transitive dependencies (induced by constraints) of the
+    /// variables in the context.
+    fn search_variable_deps(&self) -> HashSet<Meta> {
+        let mut seen = HashSet::new();
+        let mut new_variables: HashSet<Meta> = self.variables.clone();
+        while !new_variables.is_empty() {
+            let mut newer_variables = HashSet::new();
+            for m in new_variables.into_iter() {
+                if !seen.contains(&m) {
```

Comment:
You could do `if !seen.insert(m)` here (it returns a bool saying whether anything was added or not), rather than having a separate `insert` afterwards